### PR TITLE
Refactor: Implement external service adapters

### PR DESCRIPTION
**Pull request by rmulligan**
## Summary
- Implemented adapter pattern for all external service interactions
- Refactored all service classes to use adapters with dependency injection
- Added comprehensive tests for all adapters and updated services
- Enhanced services with new capabilities and improved error handling

## Implementation Details
This PR continues our refactoring effort to improve the architecture of InkLink by applying SOLID principles. Key changes include:

1. **Adapter Pattern**: Created adapter classes for all external service interactions, including:
   - HTTPAdapter: For web requests with retry logic and error handling
   - PDFAdapter: For PDF processing operations
   - GoogleAPIAdapter: For Google API interactions
   - HandwritingAdapter: For handwriting recognition
   - AIAdapter: For AI model interactions
   - QRCodeAdapter: For QR code generation
   - RemarkableAdapter: For reMarkable cloud interactions

2. **Dependency Injection**: Updated services to accept adapters through constructors with sensible defaults, making them more testable and flexible.

3. **Enhanced Testing**: Created comprehensive tests for all adapters and services, improving code coverage and reliability.

4. **New Capabilities**: Added new features like SVG QR code generation, multiple AI provider support, and enhanced PDF processing.

## Test Plan
- Run the existing test suite to verify all tests pass
- Manually verify round trip workflow with the refactored services
- Check adapter initialization and error handling with edge cases

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Comment by rmulligan** [comment]
Thanks for the review @sourcery-ai\! I've addressed the test failures by:

1. Added proper HTML content extraction utilities in 
2. Fixed the missing  function in 
3. Added missing  import in 
4. Fixed test expectations in  to work with the new adapter pattern

All tests are now passing. This PR completes the refactoring of all external service interactions to the adapter pattern, which significantly improves separation of concerns and testability throughout the codebase.

File: src/inklink/services/pdf_service.py [hunk 1]
```diff
@@ -50,15 +57,27 @@ def is_pdf_url(self, url: str) -> bool:
...
-            return "application/pdf" in content_type
+            success, headers = self.http_adapter.get(
+                url, 
+                headers={"Range": "bytes=0-0"}  # Only request header, not content
+            )
+            
+            if not success:
+                return False
+                
+            # If the response is a dict (JSON), check Content-Type header
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/services/pdf_service.py:67]: @sourcery-ai create issue

File: src/inklink/services/web_scraper_service.py [hunk 2]
```diff
@@ -98,44 +112,42 @@ def scrape(self, url: str) -> Dict[str, Any]:
...
-                "images": [],
-            }
-
-    def _fetch_url(self, url: str) -> str:
-        """Fetch URL and return its HTML content.
-
-        Uses retry logic for resilience against temporary network failures.
+            return self._build_error_response(url, f"Could not extract content: {str(e)}")
+
+    def _fetch_url(self, url: str) -> tuple[bool, str]:
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/services/web_scraper_service.py:120]: @sourcery-ai create issue

File: src/inklink/services/pdf_service.py [hunk 3]
```diff
@@ -78,72 +97,75 @@
...
+            
+        Returns:
+            List of paths to generated images
+        """
+        return self.pdf_adapter.convert_to_images(
+            pdf_path,
+            output_dir=output_dir or self.extract_dir
+        )
 
     def generate_index_notebook(
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/services/pdf_service.py:166]: @sourcery-ai create issue

File: src/inklink/adapters/pdf_adapter.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+from typing import Dict, Any, Optional, List, Tuple, BinaryIO, Union
+import PyPDF2
+from pdf2image import convert_from_path
+
+from inklink.adapters.adapter import Adapter
+
+logger = logging.getLogger(__name__)
+
+
+class PDFAdapter(Adapter):
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/adapters/pdf_adapter.py:18]: @sourcery-ai create issue

File: src/inklink/adapters/ai_adapter.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+import requests
+from typing import Dict, Any, List, Optional, Union, Tuple
+
+from inklink.adapters.adapter import Adapter
+from inklink.utils import retry_operation, format_error
+
+logger = logging.getLogger(__name__)
+
+
+class AIAdapter(Adapter):
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/adapters/ai_adapter.py:17]: @sourcery-ai create issue

File: src/inklink/adapters/google_adapter.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+import re
+import tempfile
+from urllib.parse import urlparse
+
+from inklink.adapters.adapter import Adapter
+
+logger = logging.getLogger(__name__)
+
+
+class GoogleAPIAdapter(Adapter):
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/adapters/google_adapter.py:18]: @sourcery-ai create issue

File: src/inklink/adapters/http_adapter.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+import requests
+from requests.adapters import HTTPAdapter as RequestsHTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+
+from inklink.adapters.adapter import Adapter
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPAdapter(Adapter):
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/adapters/http_adapter.py:19]: @sourcery-ai create issue

File: src/inklink/services/converters/markdown_converter.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        Returns:
+            Path to generated .rm file or None if failed
+        """
+        try:
+            url = content.get("url", "")
+            qr_path = content.get("qr_path", "")
+            title = content.get("title", f"Page from {url}")
+            
+            # Generate markdown file path if not provided
+            if not output_path:
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/services/converters/markdown_converter.py:49]: @sourcery-ai create issue

File: src/inklink/services/converters/html_converter.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+
+class HTMLConverter(BaseConverter):
+    """Converts HTML content directly to reMarkable format."""
+    
+    def can_convert(self, content_type: str) -> bool:
+        """Check if this converter can handle the given content type."""
+        return content_type == "html"
+    
+    def convert(self, content: Dict[str, Any], output_path: Optional[str] = None) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/services/converters/html_converter.py]: @sourcery-ai create issue

File: src/inklink/services/converters/markdown_converter.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+                    return None
+            else:
+                logger.error("RCU not available for Markdown conversion")
+                return None
+                
+        except Exception as e:
+            logger.error(f"Error converting markdown: {str(e)}")
+            return None
+    
+    def _write_markdown_file(self, md_path: str, url: str, qr_path: str, title: str,
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/services/converters/markdown_converter.py]: @sourcery-ai create issue

File: tests/test_google_docs_service.py [hunk 11]
```diff
@@ -1,111 +1,341 @@
...
 import os
 import pytest
+from unittest.mock import MagicMock, patch
+from typing import Dict, Any, List, Tuple
 
 from inklink.services.google_docs_service import GoogleDocsService
+from inklink.adapters.google_adapter import GoogleAPIAdapter
+
+
+class MockGoogleAPIAdapter:
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_google_docs_service.py:12]: @sourcery-ai create issue

File: tests/test_pdf_service.py [hunk 12]
```diff
@@ -1,105 +1,247 @@
...
+def test_convert_to_images(pdf_service, mock_pdf_adapter):
+    """Test PDF to image conversion."""
+    pdf_path = "/path/to/test.pdf"
+    
+    # Call the service method
+    image_paths = pdf_service.convert_to_images(pdf_path)
+    
+    # Verify adapter was called correctly
+    mock_pdf_adapter.convert_to_images.assert_called_once_with(
+        pdf_path,
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_pdf_service.py:159]: @sourcery-ai create issue

File: tests/test_handwriting_recognition.py [hunk 13]
```diff
@@ -74,79 +167,174 @@
...
     iink_data = service.convert_to_iink_format(mock_strokes)
     result = service.recognize_handwriting(iink_data)
 
+    # Verify the adapter was called
+    mock_handwriting_adapter.recognize_handwriting.assert_called_once_with(iink_data, "Text", "en_US")
+    
+    # Verify the result structure
     assert result["success"] is True
     assert result["content_id"] == "test_content_id"
     assert "raw_result" in result
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_handwriting_recognition.py:215]: @sourcery-ai create issue

File: tests/test_handwriting_recognition.py [hunk 14]
```diff
@@ -74,79 +167,174 @@ def test_extract_strokes_empty_file(tmp_path):
...
+    
+    result = service.initialize_iink_sdk("new_app_key", "new_hmac_key")
+    
+    # Verify the adapter was called
+    mock_handwriting_adapter.initialize_sdk.assert_called_once_with("new_app_key", "new_hmac_key")
+    
+    # Verify the result
+    assert result is True
+    assert service.application_key == "new_app_key"
+    assert service.hmac_key == "new_hmac_key"
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_handwriting_recognition.py:195]: @sourcery-ai create issue

File: tests/test_qr_service.py [hunk 15]
```diff
@@ -1,11 +1,261 @@
...
+def test_adapter_ping(tmp_path):
+    """Test the adapter ping functionality."""
+    with patch.object(QRCodeAdapter, 'ping', return_value=True):
+        adapter = QRCodeAdapter(output_dir=str(tmp_path))
+        assert adapter.ping() is True
+        
+    # Create a new adapter that will fail the ping
+    with patch.object(QRCodeAdapter, 'ping', return_value=False):
+        adapter = QRCodeAdapter(output_dir=str(tmp_path))
+        assert adapter.ping() is False
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_qr_service.py:219]: @sourcery-ai create issue

File: tests/test_web_scraper_service.py [hunk 16]
```diff
@@ -30,35 +64,191 @@ def test_scrape_basic(monkeypatch):
...
+    # Check the error response
     assert result["title"] == "http://example.com"
-    assert any(
-        "Could not fetch content" in item.get("content", "")
-        for item in result["structured_content"]
-    )
+    assert len(result["structured_content"]) == 1
+    assert result["structured_content"][0]["type"] == "paragraph"
+    assert "Could not fetch content: Connection error" in result["structured_content"][0]["content"]
+    assert result["images"] == []
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_web_scraper_service.py:200]: @sourcery-ai create issue

File: tests/test_qr_service.py [hunk 17]
```diff
@@ -1,11 +1,261 @@
...
+    assert len(mock_adapter.generate_qr_code_calls) == 1
+    call = mock_adapter.generate_qr_code_calls[0]
+    assert call["data"] == url
+    assert call["fill_color"] == "blue"
+    assert call["back_color"] == "yellow"
+    assert call["custom_config"]["version"] == 2
+    assert call["custom_config"]["box_size"] == 15
+    assert call["custom_config"]["border"] == 3
+    assert "fill_color" not in call["custom_config"]
+    assert "back_color" not in call["custom_config"]
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_qr_service.py:207]: @sourcery-ai create issue

File: tests/test_web_scraper_service.py [hunk 18]
```diff
@@ -30,35 +64,191 @@
...
+    
+    # Verify readability was used
+    mock_document_class.assert_called_once()
+    mock_document.summary.assert_called_once()
+    mock_document.short_title.assert_called_once()
+    
+    # Check the extracted content
+    assert result["title"] == "Clean Title"
+    assert any(item["type"] == "h2" for item in result["structured_content"])
+    assert any(item["type"] == "paragraph" for item in result["structured_content"])
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_web_scraper_service.py:181]: @sourcery-ai create issue

File: tests/test_ai_service.py [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+    response = ai_service.process_query("What is machine learning?")
+    
+    # Check empty response on failure
+    assert response == ""
+    
+    # Verify adapter was still called
+    assert len(mock_adapter.generate_structured_completion_calls) == 1
+
+
+def test_adapter_initialization():
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_ai_service.py:231]: @sourcery-ai create issue

File: src/inklink/di/service_provider.py [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+                    logger.error(f"Cannot resolve dependency {param_name} of type {param.annotation}")
+                    raise Exception(f"Cannot resolve dependency {param_name} of type {param.annotation}")
+            
+            # If parameter has a default value, use it
+            if param.default != inspect.Parameter.empty:
+                continue
+                
+            # If we got here, we couldn't figure out how to satisfy this parameter
+            logger.error(f"Cannot resolve parameter {param_name} for {class_type.__name__}")
+            raise Exception(f"Cannot resolve parameter {param_name} for {class_type.__name__}")
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/di/service_provider.py]: @sourcery-ai create issue

File: src/inklink/di/service_provider.py [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+                try:
+                    kwargs[param_name] = self.resolve(param.annotation)
+                    continue
+                except KeyError:
+                    # If we can't resolve it but it has a default value, use default
+                    if param.default != inspect.Parameter.empty:
+                        continue
+                    # Otherwise, we can't satisfy this dependency
+                    logger.error(f"Cannot resolve dependency {param_name} of type {param.annotation}")
+                    raise Exception(f"Cannot resolve dependency {param_name} of type {param.annotation}")
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply src/inklink/di/service_provider.py]: @sourcery-ai create issue

File: tests/test_google_docs_service.py [hunk 22]
```diff
@@ -1,111 +1,341 @@
...
+        "https://docs.google.com/document/d/abc123/edit",
+        "https://docs.google.com/document/d/abc123/",
+        "https://docs.google.com/document/d/abc123",
+        "https://docs.google.com/document/d/e/abc123/edit",
+    ]
+    
+    for url in url_cases:
+        doc_id = google_docs_service.adapter.extract_doc_id(url)
+        assert doc_id == "abc123"
+        assert mock_adapter.last_extracted_id == "abc123"
```
[the remark this answers is not in the document]
Comment (rmulligan) [reply tests/test_google_docs_service.py]: @sourcery-ai create issue

**Comment by rmulligan** [comment]
Thanks for the excellent architectural feedback @sourcery-ai\! I've addressed all the points you raised:

1. Created a new `RMAPIAdapter` class that encapsulates all subprocess interactions with rmapi, and refactored `AuthController` to use this adapter instead of calling subprocess directly.

2. Modified components to accept configuration via constructor parameters instead of accessing global CONFIG:
   - Updated `HCLRenderer` to accept a config dictionary 
   - Enhanced `BaseConverter` with proper configuration injection
   - Improved `PDFConverter` to inherit configuration properly
   - Updated `hcl_render.py` utilities to accept optional config

3. Improved encapsulation by making components more self-contained and consistent with the adapter pattern.

These changes have significantly improved the architectural consistency and testability throughout the codebase.

**Comment by rmulligan** [comment]
I've now addressed all the issues raised in the reviews:

1. Fixed CodeQL issues:
   - Replaced string inclusion checks with proper URL parsing in test_ai_service.py
   - Using  to properly validate domains instead of simple string containment checks

2. Addressed Sourcery architectural suggestions:
   - Created RMAPIAdapter to encapsulate subprocess calls in AuthController
   - Implemented proper configuration injection in multiple components:
     - HCLRenderer now accepts configuration via constructor
     - BaseConverter accepts and passes configuration parameters
     - Utility functions accept optional configuration

3. Fixed all flake8/black style issues:
   - Fixed bare except in test_server.py
   - Added proper spacing before inline comments
   - Added newlines at end of files
   - Properly formatted all files with black

4. Resolved merge conflicts with main branch

All tests are now passing and all checks should be green. The code now consistently follows the adapter pattern throughout the codebase and properly implements dependency injection for configuration.

**Comment by rmulligan** [comment]
@sourcery-ai resolve
